// @qlik/sdk/dist/cjs/qlik.js
var __create = Object.create;
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __getProtoOf = Object.getPrototypeOf;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toESM = (mod, isNodeMode, target) => (target = mod != null ? __create(__getProtoOf(mod)) : {}, __copyProps(
  isNodeMode || !mod || !mod.__esModule ? __defProp(target, "default", { value: mod, enumerable: true }) : target,
  mod
));
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var qlik_exports = {};
__export(qlik_exports, {
  ApiKeys: () => import_Api_Keys2.ApiKeys,
  Apps: () => import_Apps2.Apps,
  Auth: () => import_auth.default,
  AuthType: () => import_types2.AuthType,
  AuthenticationErrorMsg: () => import_errors.AuthenticationErrorMsg,
  Automations: () => import_Automations2.Automations,
  Collections: () => import_Collections2.Collections,
  ConnectionErrorMsg: () => import_errors.ConnectionErrorMsg,
  CspOrigins: () => import_Csp_Origins2.CspOrigins,
  CustomErrorMsg: () => import_errors.CustomErrorMsg,
  DataFiles: () => import_Data_Files2.DataFiles,
  Extensions: () => import_Extensions2.Extensions,
  Groups: () => import_Groups2.Groups,
  IdentityProviders: () => import_Identity_Providers2.IdentityProviders,
  Items: () => import_Items2.Items,
  Licenses: () => import_Licenses2.Licenses,
  OauthTokens: () => import_Oauth_Tokens2.OauthTokens,
  Qix: () => Qix,
  Quotas: () => import_Quotas2.Quotas,
  ReloadTasks: () => import_Reload_Tasks2.ReloadTasks,
  Reloads: () => import_Reloads2.Reloads,
  Roles: () => import_Roles2.Roles,
  Spaces: () => import_Spaces2.Spaces,
  Themes: () => import_Themes2.Themes,
  Users: () => import_Users2.Users,
  WebIntegrations: () => import_Web_Integrations2.WebIntegrations,
  Webhooks: () => import_Webhooks2.Webhooks,
  default: () => Qlik
});
module.exports = __toCommonJS(qlik_exports);
var import_auth = __toESM(require("./auth/auth.js"));
var import_Extensions = require("./apis/Extensions.js");
var import_Themes = require("./apis/Themes.js");
var import_Apps = require("./apis/Apps.js");
var import_Items = require("./apis/Items.js");
var import_Spaces = require("./apis/Spaces.js");
var import_Users = require("./apis/Users.js");
var import_Csp_Origins = require("./apis/Csp_Origins.js");
var import_Reloads = require("./apis/Reloads.js");
var import_Data_Files = require("./apis/Data_Files.js");
var import_Web_Integrations = require("./apis/Web_Integrations.js");
var import_Roles = require("./apis/Roles.js");
var import_Licenses = require("./apis/Licenses.js");
var import_Groups = require("./apis/Groups.js");
var import_Identity_Providers = require("./apis/Identity_Providers.js");
var import_Api_Keys = require("./apis/Api_Keys.js");
var import_Quotas = require("./apis/Quotas.js");
var import_Webhooks = require("./apis/Webhooks.js");
var import_Collections = require("./apis/Collections.js");
var import_Automations = require("./apis/Automations.js");
var import_Reload_Tasks = require("./apis/Reload_Tasks.js");
var import_Oauth_Tokens = require("./apis/Oauth_Tokens.js");
var Qix = __toESM(require("./apis/Qix.js"));
var import_types2 = require("./types/types.js");
var import_errors = require("./errors.js");
var import_Extensions2 = require("./apis/Extensions.js");
var import_Themes2 = require("./apis/Themes.js");
var import_Apps2 = require("./apis/Apps.js");
var import_Items2 = require("./apis/Items.js");
var import_Spaces2 = require("./apis/Spaces.js");
var import_Users2 = require("./apis/Users.js");
var import_Csp_Origins2 = require("./apis/Csp_Origins.js");
var import_Reloads2 = require("./apis/Reloads.js");
var import_Data_Files2 = require("./apis/Data_Files.js");
var import_Web_Integrations2 = require("./apis/Web_Integrations.js");
var import_Roles2 = require("./apis/Roles.js");
var import_Licenses2 = require("./apis/Licenses.js");
var import_Groups2 = require("./apis/Groups.js");
var import_Identity_Providers2 = require("./apis/Identity_Providers.js");
var import_Api_Keys2 = require("./apis/Api_Keys.js");
var import_Quotas2 = require("./apis/Quotas.js");
var import_Webhooks2 = require("./apis/Webhooks.js");
var import_Collections2 = require("./apis/Collections.js");
var import_Automations2 = require("./apis/Automations.js");
var import_Reload_Tasks2 = require("./apis/Reload_Tasks.js");
var import_Oauth_Tokens2 = require("./apis/Oauth_Tokens.js");
class Qlik {
  constructor(config) {
    this.auth = new import_auth.default(config);
    this.rest = this.auth.rest;
    this.rpc = this.auth.rpc;
    this.apps = new import_Apps.Apps(this.auth);
    this.users = new import_Users.Users(this.auth);
    this.extensions = new import_Extensions.Extensions(this.auth);
    this.themes = new import_Themes.Themes(this.auth);
    this.spaces = new import_Spaces.Spaces(this.auth);
    this.items = new import_Items.Items(this.auth);
    this.csporigins = new import_Csp_Origins.CspOrigins(this.auth);
    this.reloads = new import_Reloads.Reloads(this.auth);
    this.datafiles = new import_Data_Files.DataFiles(this.auth);
    this.webintegrations = new import_Web_Integrations.WebIntegrations(this.auth);
    this.roles = new import_Roles.Roles(this.auth);
    this.licenses = new import_Licenses.Licenses(this.auth);
    this.groups = new import_Groups.Groups(this.auth);
    this.identityProviders = new import_Identity_Providers.IdentityProviders(this.auth);
    this.apiKeys = new import_Api_Keys.ApiKeys(this.auth);
    this.quotas = new import_Quotas.Quotas(this.auth);
    this.webhooks = new import_Webhooks.Webhooks(this.auth);
    this.collections = new import_Collections.Collections(this.auth);
    this.automations = new import_Automations.Automations(this.auth);
    this.reloadTasks = new import_Reload_Tasks.ReloadTasks(this.auth);
    this.oauthTokens = new import_Oauth_Tokens.OauthTokens(this.auth);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  ApiKeys,
  Apps,
  Auth,
  AuthType,
  AuthenticationErrorMsg,
  Automations,
  Collections,
  ConnectionErrorMsg,
  CspOrigins,
  CustomErrorMsg,
  DataFiles,
  Extensions,
  Groups,
  IdentityProviders,
  Items,
  Licenses,
  OauthTokens,
  Qix,
  Quotas,
  ReloadTasks,
  Reloads,
  Roles,
  Spaces,
  Themes,
  Users,
  WebIntegrations,
  Webhooks
});
