// @qlik/sdk/dist/esm/apis/Themes.js
import Auth from "../auth/auth.js";
class Theme {
  constructor({
    author,
    createdAt,
    dependencies,
    description,
    file,
    homepage,
    icon,
    id,
    keywords,
    license,
    name,
    qextFilename,
    qextVersion,
    repository,
    supplier,
    tags,
    tenantId,
    type,
    updateAt,
    userId,
    version,
    ...rest
  }, auth) {
    this.auth = auth;
    this.author = author;
    this.createdAt = createdAt;
    this.dependencies = dependencies;
    this.description = description;
    this.file = file;
    this.homepage = homepage;
    this.icon = icon;
    this.id = id;
    this.keywords = keywords;
    this.license = license;
    this.name = name;
    this.qextFilename = qextFilename;
    this.qextVersion = qextVersion;
    this.repository = repository;
    this.supplier = supplier;
    this.tags = tags;
    this.tenantId = tenantId;
    this.type = type;
    this.updateAt = updateAt;
    this.userId = userId;
    this.version = version;
    if (rest) {
      Object.keys(rest).forEach((prop) => {
        this[prop] = rest[prop];
      });
    }
  }
  async getFile() {
    await this.auth.rest(
      "/themes/{id}/file".replace("{id}", this.id ?? ""),
      {
        method: "GET"
      }
    );
  }
  async delete() {
    await this.auth.rest(
      "/themes/{id}".replace("{id}", this.id ?? ""),
      {
        method: "DELETE"
      }
    );
  }
  async patches(data, file) {
    const formData = new FormData();
    if (file !== void 0) {
      formData.append("file", new Blob([file], { type: "application/octet-stream" }));
    }
    if (data !== void 0) {
      formData.append("data", JSON.stringify(data));
    }
    const response = await this.auth.rest(
      "/themes/{id}".replace("{id}", this.id ?? ""),
      {
        method: "PATCH",
        body: formData
      }
    );
    const responseJson = await response.json();
    return new ThemesClass(responseJson, this.auth);
  }
}
class ThemeDef {
  constructor({
    author,
    dependencies,
    description,
    file,
    homepage,
    icon,
    keywords,
    license,
    name,
    qextFilename,
    qextVersion,
    repository,
    supplier,
    tags,
    type,
    version,
    ...rest
  }, auth) {
    this.auth = auth;
    this.author = author;
    this.dependencies = dependencies;
    this.description = description;
    this.file = file;
    this.homepage = homepage;
    this.icon = icon;
    this.keywords = keywords;
    this.license = license;
    this.name = name;
    this.qextFilename = qextFilename;
    this.qextVersion = qextVersion;
    this.repository = repository;
    this.supplier = supplier;
    this.tags = tags;
    this.type = type;
    this.version = version;
    if (rest) {
      Object.keys(rest).forEach((prop) => {
        this[prop] = rest[prop];
      });
    }
  }
}
class ThemesClass {
  constructor({
    data,
    ...rest
  }, auth) {
    this.auth = auth;
    if (data !== void 0 && data !== null) {
      if (data.length > 0 && data[0] instanceof Theme) {
        this.data = data;
      } else {
        this.data = data.map((element) => new Theme(element, this.auth));
      }
    }
    if (rest) {
      Object.keys(rest).forEach((prop) => {
        this[prop] = rest[prop];
      });
    }
  }
}
class Themes {
  constructor(config) {
    this.config = config instanceof Auth ? config.config : config;
    this.auth = config instanceof Auth ? config : new Auth(config);
  }
  async getFile(filepath, id) {
    await this.auth.rest(
      "/themes/{id}/file/{filepath}".replace("{filepath}", filepath).replace("{id}", id),
      {
        method: "GET"
      }
    );
  }
  async get(id) {
    const response = await this.auth.rest(
      "/themes/{id}".replace("{id}", id),
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new Theme(responseJson, this.auth);
  }
  async getThemes() {
    const response = await this.auth.rest(
      "/themes",
      {
        method: "GET"
      }
    );
    const responseJson = await response.json();
    return new ThemesClass(responseJson, this.auth);
  }
  async create(data, file) {
    const formData = new FormData();
    if (file !== void 0) {
      formData.append("file", new Blob([file], { type: "application/octet-stream" }));
    }
    if (data !== void 0) {
      formData.append("data", JSON.stringify(data));
    }
    const response = await this.auth.rest(
      "/themes",
      {
        method: "POST",
        body: formData
      }
    );
    const responseJson = await response.json();
    return new Theme(responseJson, this.auth);
  }
}
export {
  Theme,
  ThemeDef,
  Themes,
  ThemesClass
};
