// @qlik/sdk/dist/cjs/types/types.js
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var types_exports = {};
__export(types_exports, {
  AuthType: () => AuthType
});
module.exports = __toCommonJS(types_exports);
var AuthType = /* @__PURE__ */ ((AuthType2) => {
  AuthType2["APIKey"] = "APIKey";
  AuthType2["WebIntegration"] = "WebIntegration";
  AuthType2["JWTAuth"] = "JWTAuth";
  AuthType2["OAuth2"] = "OAuth2";
  AuthType2["None"] = "None";
  return AuthType2;
})(AuthType || {});
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AuthType
});
